sdk-tools/src/edger8r.rs

```diff
@@ -0,0 +1,374 @@
+    edger8r_path: PathBuf,
+    /// The build output directory.
+    out_dir: Option<PathBuf>,
+    /// The path to the EDL file to generate code for.
+    edl_file: PathBuf,
+    /// The type of code to be generated.
+    output_kind: OutputKind,
+}
+
+/// The files generated by the edger8r tool
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Edger8rFiles {
+    /// The trusted files generated. Will be empty when
+    /// [`OutputKind::Untrusted`] is used.
+    pub trusted: Vec<PathBuf>,
+    /// The untrusted files generated. Will be empty when
+    /// [`OutputKind::Trusted`] is used.
+    pub untrusted: Vec<PathBuf>,
+}
+
+impl Edger8r {
+    /// Create a new edger8r builder.
+    ///
+    /// # Arguments
+    /// * `edl_file` - The edl file to generate C files from
+    pub fn new(edl_file: impl AsRef<Path>) -> Self {
+        let edl_file = edl_file.as_ref().to_owned();
+
+        Self {
+            edger8r_path: Self::edger8r_path(),
+            out_dir: None,
+            edl_file,
+            output_kind: OutputKind::Both,
+        }
+    }
+
+    /// Set the type of code we should generate
+    ///
+    /// # Arguments
+    /// * `kind` - The kind of code to generate
+    pub fn output_kind(&mut self, kind: OutputKind) -> &mut Self {
+        self.output_kind = kind;
+        self
+    }
+
+    /// Specify the output directory
+    ///
+    /// # Arguments
+    /// * `out_dir` - the output directory to place the generated files in.
+    pub fn out_dir(&mut self, out_dir: impl AsRef<Path>) -> &mut Self {
+        self.out_dir = Some(out_dir.as_ref().to_owned());
+        self
+    }
+
+    /// Generate C files using the edger8r tool
+    ///
+    /// # Returns:
+    /// The paths to the generated files.
+    pub fn generate(&self) -> Result<Edger8rFiles, Error> {
+        let mut command = Command::new(&self.edger8r_path);
+
+        let mut dirs = vec![];
+        match self.output_kind {
+            OutputKind::Trusted => {
+                command.arg("--trusted");
+                dirs.push("--trusted-dir");
+            }
+            OutputKind::Untrusted => {
+                command.arg("--untrusted");
+                dirs.push("--untrusted-dir");
+            }
+            OutputKind::Both => {
+                dirs.extend(["--trusted-dir", "--untrusted-dir"]);
+            }
+        }
+
+        let out_dir = match self.out_dir.as_ref() {
+            None => Self::default_out_dir(),
+            Some(dir) => dir.clone(),
+        };
+
+        command.arg(&self.edl_file);
+        for d in dirs {
+            command.arg(d).arg(&out_dir);
+        }
+
+        let output = command.output()?;
+
+        if output.status.success() {
+            Ok(self.output_file_names(out_dir))
+        } else {
+            Err(Error::Generate(
+                format!("{command:?}"),
+                String::from_utf8(output.stdout)?,
+                String::from_utf8(output.stderr)?,
```

Comment:
I'm thinking that perhaps these should be upgraded to `from_utf8_lossy`.  I think it unlikely that the edger8r will provide invalid utf8, but if it does it seems like we would want to provide more information to users than `Error::Utf8Error`

Reply:
Yep, agreed.

Reply:
done